File: objc/Facade/Facade/Protocol.m
#import "Protocol.h"

@implementation Protocol {
    DataCompressor* dataCompressor;
    DataSpliter* dataSpliter;
    PacketSender* packetSender;
}

-(instancetype)initWith:(CompressionAlgorithm)algo packetSize:(int)packetSize {
    if(self = [super init]) {
        dataCompressor = [[DataCompressor alloc] init];
        dataCompressor.compressionAlgo = algo;
        dataSpliter = [[DataSpliter alloc] initWithPacketSize:packetSize];
        packetSender = [[PacketSender alloc] init];
    }
    return self;
}

-(void)sendData:(NSString*)data {
    data = [dataCompressor compress:data];
    NSArray* packetList = [dataSpliter splitData:data];
    for(NSString* packet in packetList)
        [packetSender sendPacket:packet];
}

@end

File: swift/Facade/Facade/Protocol.swift
import Foundation

enum CompressionAlgorithm {
    case CompressionAlgorithm_GZIP
    case CompressionAlgorithm_BZIP
}

class Protocol {
    private let dataCompressor: DataCompressor
    private let dataSpliter: DataSpliter
    private let packetSender: PacketSender
    
    init(_ algo: CompressionAlgorithm, _ packetSize: Int) {
        self.dataCompressor = DataCompressor(algo)
        self.dataSpliter = DataSpliter(packetSize)
        self.packetSender = PacketSender()
    }
    
    func sendData(_ data: String) -> Void {
        let compressedData: String? = dataCompressor.compress(data)
        let packetList: [String] = dataSpliter.splitData(compressedData!)
        for packet in packetList {
            packetSender.sendPacket(packet)
        }
    }
}
